File: OOP/Inheritance.js
class Animal {
  constructor(name) { this.name = name; }
  speak() { return `${this.name} makes a noise.`; }
}

class Dog extends Animal {
  constructor(name, breed) {
    super(name);          // calls parent constructor
    this.breed = breed;
  }

  speak() {    // Method overriding
    return `${this.name} barks.`;
  }

  info() {
    return `${super.speak()} — ${this.breed}`;  // calls parent method
  }
}

const d = new Dog("Rex", "Labrador");
console.log(d.speak());  // "Rex barks."
console.log(d.info());   // "Rex makes a noise. — Labrador"

File: OOP/Inheritance.py
class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        return f"{self.name} makes a noise."


class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name)
        self.breed = breed

    def speak(self):
        return f"{self.name} barks."

    def info(self):
        return f"{super().speak()} - {self.breed}"


d = Dog("Rex", "Labrador")
print(d.speak())
print(d.info())
